### PR TITLE
Fix BlockingIOError as a result of too large data passed to strulogs

### PR DESCRIPTION
**Pull request by Tawakalt**
**Proposed changes**:
- Fixes [ATO-1200](https://rasahq.atlassian.net/browse/ATO-1200)

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [X] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

[ATO-1200]: https://rasahq.atlassian.net/browse/ATO-1200?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rasa/core/brokers/kafka.py [hunk 1]
```diff
@@ -240,6 +241,9 @@ def _publish(self, event: Dict[Text, Any]) -> None:
                 )
             ]
 
+        # set stdout to non-blocking
+        fcntl.fcntl(1, fcntl.F_SETFL, 0)
```
Comment (ancalita) [inline rasa/core/brokers/kafka.py:245]: I think this won't work on Windows, it's only a solution for Unix systems, we'll need something universal for any OS.

Comment (ancalita) [reply rasa/core/brokers/kafka.py:245]: Also, we should probably seek all potential log locations with too much data that causes this issue, not only this module.

Comment (Tawakalt) [reply rasa/core/brokers/kafka.py:245]: Yeah, tests are already failing.

Comment (Tawakalt) [reply rasa/core/brokers/kafka.py:245]: > Also, we should probably seek all potential log locations with too much data that causes this issue, not only this module.
Only this particular log caused the issue. But I can future proof all logs with large data.

Comment (ancalita) [reply rasa/core/brokers/kafka.py:245]: 🙏🏻

Comment (Tawakalt) [reply rasa/core/brokers/kafka.py:245]: > replace keys that could hold too much data with an optimized version of that, e.g. rasa_event, rasa_events, extraction_events etc.

Are we saying we can send out a reduced version instead? Like with less data?

Comment (ancalita) [reply rasa/core/brokers/kafka.py:245]: Something worth exploring based on how previous logs used to look like.

File: tests/shared/core/test_events.py [hunk 2]
```diff
@@ -801,3 +801,29 @@ def test_event_fingerprint_uniqueness(event: Event):
...
+            "name": "action_listen",
+            "policy": None,
+            "confidence": None,
+            "timestamp": None,
+        },
+    ],
+)
+def test_reduce_event(event: Dict[Text, Any]):
+    reduced_event = rasa.shared.core.events.reduce_event(event)
+    assert "parse_data" not in reduced_event
```
Comment (Tawakalt) [inline tests/shared/core/test_events.py:829]: I confirmed that only events containing `parse_data` i.e `user events` have very large data that'll trigger the error.

File: rasa/shared/core/events.py [hunk 3]
```diff
@@ -246,6 +246,21 @@ def do_events_begin_with_session_start(events: List["Event"]) -> bool:
...
+
+    Args:
+        event: The event to be reduced.
+
+    Returns:
+        A reduced copy of the event.
+    """
+    reduced_event = copy.deepcopy(event)
+    if "parse_data" in reduced_event:
+        del reduced_event["parse_data"]
```
Comment (Urkem) [inline rasa/shared/core/events.py:260]: Was there any particular reason to just delete `parse_data` and not summarise it? (maybe just the first 1-3 results)

File: rasa/shared/core/events.py [hunk 4]
```diff
@@ -246,6 +246,21 @@ def do_events_begin_with_session_start(events: List["Event"]) -> bool:
     )
 
 
+def reduce_event(event: Dict[Text, Any]) -> Dict[Text, Any]:
```
Comment (Urkem) [inline rasa/shared/core/events.py:249]: Probably better to name it something like `remove_parse_data`.

File: rasa/core/brokers/pika.py [hunk 5]
```diff
@@ -298,20 +298,24 @@ async def _publish(
...
-                rasa_event=copy.deepcopy(event),
+                rasa_event=copy.deepcopy(reduced_rasa_event),
             )
         except Exception as e:
             structlogger.error(
                 "pika.events.publish.failed",
+                event_info="Logging a reduced version of the failed Pika event",
                 host=self.host,
-                rasa_event=copy.deepcopy(event),
+                rasa_event=copy.deepcopy(reduced_rasa_event),
```
Comment (Urkem) [inline rasa/core/brokers/pika.py:318]: Why are we `deepcopy`-ing the event again here? `reduce_event` is it already doing the copy

**Review by Urkem — APPROVED** [review]
LGTM 👍

**Pull request merged**